### PR TITLE
callback filter, no content type sniff, prepend empty JS comment

### PR DESCRIPTION
**Pull request by quentin-**

### CODE REVIEW

File: lib/rack/jsonp.rb [hunk 1]
```diff
@@ -60,7 +62,7 @@ def pad(callback, response, body = "")
         body << s.to_s.gsub("\u2028", '\u2028').gsub("\u2029", '\u2029')
       end
       close(response)
-      ["#{callback}(#{body})"]
+      ["/**/#{callback}(#{body})"]
```
Comment (ejholmes) [inline lib/rack/jsonp.rb:65]: I'm not sure if I feel that this is incredibly clever, or to feel disgusted by browser security.

File: lib/rack/jsonp.rb [hunk 2]
```diff
@@ -12,6 +12,7 @@ def initialize(app, options = {})
       @carriage_return = options[:carriage_return] || false
       @callback_param = options[:callback_param] || 'callback'
       @timestamp_param = options[:timestamp_param] || '_'
+      @callback_regex = options[:callback_regex] || /^[a-zA-Z0-9]{1,32}$/
```
Comment (ejholmes) [inline lib/rack/jsonp.rb:15]: I think this needs to include `_`, which is used in the callback that we currently use.

**Review by ejholmes — APPROVED** [review]

File: lib/rack/jsonp.rb (hunk 2 above)
Comment (ejholmes) [reply lib/rack/jsonp.rb:15]: Ah, totally missed the internal PR for this. Feel free to ignore, this is a better default.

File: lib/rack/jsonp.rb (hunk 1 above)
Comment (quentin-) [reply lib/rack/jsonp.rb:65]: Wait untill you see why non jsonp GET xhr response should all purposely crash the JS runtime. Facebook prepends an infinite loop `for(;;;); `

File: lib/rack/jsonp.rb [hunk 3]
```diff
@@ -35,8 +36,9 @@ def call(env)
 
       status, headers, response = @app.call(env)
 
-      if callback && headers['Content-Type'] =~ /json/i
+      if callback && callback =~ @callback_regex && headers['Content-Type'] =~ /json/i
         response = pad(callback, response)
+        headers['Content-Type-Options'] = 'nosniff'
```
Comment (ejholmes) [inline lib/rack/jsonp.rb:41]: I missed this in my first review. I think this is supposed to be `X-Content-Type-Options`?

https://www.owasp.org/index.php/OWASP_Secure_Headers_Project#xcto

**Pull request merged**